Add developer setup script for smoother setup

dbalatero opened the pull request:
Fixes #2529

### Summary:

As a new Status contributor, I decided as part of my initial repo setup to try and write a repeatable Bash script to install/configure any developer dependencies. The goal is to have a new developer just run `./dev-setup.sh`, vs. following a bunch of wiki steps. I think lowering the barrier to entry is probably a good thing to do.

Curious what anyone's thoughts are on including a script like this?

# Notes

* This currently only supports OS X, but extending it for Windows/Linux would be great future enhancement. Currently it exits with a message if you're not on OS X. I don't have a test machine for other platforms at the moment.
* The script prefers using `rvm` and `nvm` for managing Ruby/Node versions, but falls back to system Ruby and `brew install node` if they aren't present.
* I am currently getting an error when running the `mvn ... dependency:unpack` script which I'd like to resolve before merging this. Here is the output I'm getting–it seems related to an error 500 being returned from ci.status.im?

```
$ mvn -f modules/react-native-status/ios/RCTStatus dependency:unpack

POM file modules/react-native-status/ios/RCTStatus specified with the -f/--file command line argument does not exist
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building Messenger 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-dependency-plugin:2.10:unpack (default-cli) @ Messenger ---
[INFO] Configured Artifact: status-im:status-go-ios-simulator:develop-gad9a877:zip
Downloading: http://139.162.11.12:8081/artifactory/libs-release-local/status-im/status-go-ios-simulator/develop-gad9a877/status-go-ios-simulator-develop-gad9a877.zip
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.937 s
[INFO] Finished at: 2017-11-27T03:10:17-08:00
[INFO] Final Memory: 17M/57M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:2.10:unpack (default-cli) on project Messenger: Unable to resolve artifact.: Could not transfer artifact status-im:status-go-ios-simulator:zip:develop-gad9a877 from/to ci.status.im (http://139.162.11.12:8081/artifactory/libs-release-local/): Failed to transfer file: http://139.162.11.12:8081/artifactory/libs-release-local/status-im/status-go-ios-simulator/develop-gad9a877/status-go-ios-simulator-develop-gad9a877.zip. Return code is: 500 , ReasonPhrase:Internal Server Error.
[ERROR]   status-im:status-go-ios-simulator:zip:develop-gad9a877
[ERROR]
[ERROR] from the specified remote repositories:
[ERROR]   ci.status.im (http://139.162.11.12:8081/artifactory/libs-release-local/, releases=true, snapshots=true),
[ERROR]   central (https://repo.maven.apache.org/maven2, releases=true, snapshots=false)
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

status: wip

Fixes #2529

pablanopete commented:
@dbalatero thanks for the contribution! Shows some real initiative, and I believe making the development setup process more smooth is certainly a worthy venture! 

I’d love for you to join the conversation at https://chat.status.im/#/register and ping me @hutch if you haven’t yet so we can chat more!

yenda commented on package-lock.json:
```diff
@@ -446,12 +446,56 @@
           "resolved": "https://registry.npmjs.org/babel-register/-/babel-register-6.26.0.tgz",
           "integrity": "sha1-btAhFz4vy0htestFxgCahW9kcHE=",
           "requires": {
+            "babel-core": "6.26.0",
```
don't modify the package.lock file

dbalatero replied:
Yeah, I thought I removed that sorry. It's pretty confusing (and IMHO wrong unless I'm misunderstanding how it should work) that `npm` will generate different package lock files on different machines without having modified any of the packages in my PR. I've found that Yarn is much more stable about its lock file, and it doesn't magically change between machines.

yenda commented:
Thank you for your contribution.

Could you please create an issue describing the problem you are trying to solve so there is a place to discuss your solution, its trade-offs, and its alternatives.

dbalatero committed:
Added developer setup script for smoother setup

Moved run-osx.sh to scripts/ subdirectory

dbalatero commented:
@yenda Totally! Here it is: https://github.com/status-im/status-react/issues/2529

oskarth commented:
@dbalatero Thanks for creating the associated issue!

Re mvn problem:

> develop-gad9a877:zip

I misdiagnosed your problem earlier (I wrote what @pablanopete copy-pasted). We had to wipe Artifactory due to running out of disk space, so this old artifact is probably no longer present. Please try pulling latest develop and you should see it download develop-g6927638 instead. See https://github.com/status-im/status-react/search?utf8=%E2%9C%93&q=develop-g6927638&type=

oskarth commented:
I'l look at it more carefully later, but just two things for now:

1) To clarify, it doesn't introduce any new dependency in the form of nvm, rvm or anything else not already required?

2) Could you put this in a `scripts` under, say, `/setup-osx`? We have a bunch of similar scripts and it'd be useful to gather them in one place with descriptive names.

dbalatero commented:
@oskarth 
> Please try pulling latest develop and you should see it download develop-g6927638 instead 

did a rebase and re-ran my setup script, and looks like I'm back in business (currently downloading the new artifact). Thanks!

> To clarify, it doesn't introduce any new dependency in the form of nvm, rvm or anything else not already required?

Correct, but it preferentially will use those tools if and only if they are installed. I personally would prefer to use those tools on my own machine.

> Could you put this in a scripts under, say, /setup-osx?

Absolutely. I'd like to clean up the functions and possibly organize them in separate files I can `source` as well, so that the main setup script reads better.

oskarth commented:
I also think it'd be useful to have a more POSIX/portable shell script so it can be run outside in eshell for example (probably without the colors). But also I realize this might be out of scope for what you are trying to achieve, since most Mac users are happy to use Terminal/iTerm with bash. This is a step in the right direction for newcomers generally, so we can always iterate on this (and have different versions possibly). Just something to consider, robust vs fancy (cecho stuff) trade-off.

dbalatero commented:
> I also think it'd be useful to have a more POSIX/portable shell script so it can be run outside in eshell for example (probably without the colors)

Happy to look into that, sounds fun. Maybe `cecho()` can just cleverly fallback to `echo` if it detects no color support.

dbalatero commented:
@oskarth Made some updates to the PR.

1. Organized functions into separate files, to keep the main script readable for anyone curious.
2. Disabled colors if the terminal doesn't support it.
3. Moved relevant root directory scripts into `scripts/`

oskarth commented:
Thanks @dbalatero! I asked some other devs to take a look at it too. I have some concerns about it being a bit "heavy" and tightly coupled around a particular style (I noticed you do this for all stuff in your dotfiles!) which might cause issues down the road. I'd consider doing a more minimal POSIX no-frills small robust one line shell script but YMMV. This is definitely better than nothing and you have a bunch of fallbacks and error handling it seems, so that's great! Approving.

oskarth commented:
I'd like to see one more approve review from this swarm https://github.com/status-im/ideas/issues/9 before we move it to testing. @yenda perhaps you guys want to actually test this PR?

dbalatero commented:
> small robust one line shell script

@oskarth I can definitely get rid of the various separate files and roll it back into one large script file if you think that might be easier. I can see pros to that.

dbalatero commented:
For testing, I'll build + spin up a Virtualbox image with OS X, and see what the experience is out of the box with a pretty fresh machine.

https://github.com/geerlingguy/macos-virtualbox-vm

dbalatero commented:
@pablanopete @andytudhope @yenda 

Thanks for the approvals all. I'll report back when I complete a test on a fresh OS X install. Please feel free to run this script on your own machines as well for testing. My goal is that devs can constantly run this script to stay up to date with the latest dependencies, e.g. "repo not working? re-run `scripts/setup-osx`"

Given that goal, I wonder if it should be called something other than "setup".

Also would like some feedback on whether it's a priority to extend setup to work on Linux and/or Windows. Not sure the % demographics of Status developers' operating systems. If it seems useful, I can start in on that after this is merged. Just let me know!

andytudhope commented:
Just currently freaking out that anyone can root my Mac with an empty password 😱 
Happy to keep it as `setup-osx`, as even the re-install stuff is still, technically, setting up.
I would say definitely a priority to extend to Linux, and maybe you can convince @Alonski to help you with the Windows stuff (commenter on actual issue for this PR), though far fewer devs use it, and none in the core team that I am aware of.

dbalatero commented:
> I would say definitely a priority to extend to Linux, and maybe you can convince @Alonski to help you with the Windows stuff

Super, should I open an issue for Linux?

andytudhope commented:
@oskarth @yenda WDYT? I reckon it's best to keep it under here so we can assign a Medium-sized bounty to this for all scripts, but might make the PR big and unwieldy... Not sure?

dbalatero commented:
(my vote would be minimum viable product-style and iterate, to keep the PR review/testing focused and small!)

dbalatero commented:
Status builds and runs on my machine, from this setup script. I also got through an entire setup on a fresh install of OS X in a VM. Still working on building.

Speaking of building, I saw this cryptic error right at the end of my build:

```
fatal: ambiguous argument 'master..': unknown revision or path not in the working tree.

Use '--' to separate paths from revisions, like this:

'git <command> [<revision>...] -- [<file>...]'

fatal: bad revision '^'

Command /bin/sh failed with exit code 128



** BUILD FAILED **


The following commands produced analyzer issues:
        Analyze RNFIRMessaging.m

(1 command with analyzer issues)

The following build commands failed:
        PhaseScriptExecution Run\ Script build/Build/Intermediates.noindex/StatusIm.build/Debug-iphonesimulator/StatusIm.build/Script-E3914A731DF919ED00EBB515.sh
(1 failure)

Installing build/Build/Products/Debug-iphonesimulator/StatusIm.app
Launching im.status.ethereum
im.status.ethereum: 43868
```

It still deployed to the iOS simulator and ran, but I was wondering if this is a known issue, or if I forgot something?

Still, it seems to work!

<img width="424" alt="screen shot 2017-11-29 at 12 25 28 am" src="https://user-images.githubusercontent.com/59429/33365149-1176e2cc-d49c-11e7-830a-c4d5fd668096.png">

dbalatero commented:
Did some more testing and I'm feeling good. What's the next step? Looks like there's a bounty-pending-approval tag on the issue, and others probably want to test the script as well.

I filed this side issue that I noticed as a result of doing this setup work as well: https://github.com/status-im/status-react/issues/2541

oskarth commented:
@andytudhope @dbalatero I agree with keeping it small, how about 1 small bounty for OSX, 1 for Linux and then we can see about Windows one? I'm not sure how useful Windows would be, to be honest. Standard advice for Windows devs if often "run Ubuntu in a VM".

oskarth commented:
@dbalatero Re `fatal: ambiguous argument 'master..': unknown revision or path not in the working tree.` please see https://github.com/status-im/status-react/issues/2531#issuecomment-347349409

You can also see this in our CI setup file: https://github.com/status-im/status-react/blob/develop/Jenkinsfile#L17-L20

andytudhope commented:
OK, am testing this now (though struggling with a bad line). Next steps are here for you @dbalatero: https://github.com/status-im/status-react/issues/2548

If you could sign in to openbounty.status.im and update your payment details with an Ethereum address you own, that'd also be helpful. We do want to assign a small bounty to this and to the Linux issue, so there's incentive to get that done soon too 😉

andytudhope commented:
@dbalatero any ideas about this?
```
$ cd ios && pod install && cd ..

Analyzing dependencies
[!] The version of CocoaPods used to generate the lockfile (1.3.1) is higher than the version of the current executable (1.1.1). Incompatibility issues may arise.
[!] Unable to satisfy the following requirements:

- `Instabug (~> 7.0)` required by `Podfile`
- `Instabug (= 7.2.6)` required by `Podfile.lock`

None of your spec sources contain a spec satisfying the dependencies: `Instabug (~> 7.0), Instabug (= 7.2.6)`.

You have either:
 * out-of-date source repos which you can update with `pod repo update`.
 * mistyped the name or version.
 * not added the source repo that hosts the Podspec to your Podfile.

Note: as of CocoaPods 1.0, `pod repo update` does not happen on `pod install` by default.
```

Otherwise, LGTM.

dbalatero commented:
> Re fatal: ambiguous argument 'master..': unknown revision or path not in the working tree. please see #2531 (comment)

@oskarth got it. Do we have control over whatever process is trying to compare `master` in the build step? Maybe we can just change it to `develop..` if so?

> OK, am testing this now (though struggling with a bad line). 

@andytudhope Interesting. It looks like you have cocoapods 1.1.1 installed, but the latest `Podfile.lock` was build with 1.3.1. My guess is that the Podfile has evolved over time but you haven't updated either `gem install cocoapods` and/or ran `pod install` since?

I will update the script to check your cocoapods version against the committed lockfile version, and swap out your versions if they mismatch. I'd like this setup script to fix this issue for anyone else as well.

dbalatero commented:
> If you could sign in to openbounty.status.im and update your payment details with an Ethereum address you own, that'd also be helpful.

Done.

oskarth commented:
> @oskarth got it. Do we have control over whatever process is trying to compare master in the build step? Maybe we can just change it to develop.. if so?

Not that I know of but possibly. This seems like a separate issue though. Feel free to open an issue and/or PR if you have a way to solve this :) (https://github.com/status-im/status-react/blob/develop/Jenkinsfile#L17-L20) - doesn't strike me as high prio but it'd be a nice to have to streamline some stuff.

dbalatero commented:
@andytudhope I rebased `develop`, and added some code that ensures you're on whatever cocoapods version that the lock file is built with. Since I did `git push --force-with-lease` would you reset your local checkout of this branch and re-run the script?

Here's the relevant output when I ran the updated script (I installed v 1.1.0 first and reproduced your issue):

```
~/code/status-react (experiment) ☠  [2.4.1] [stable] % gem install cocoapods -v 1.1.0
Fetching: cocoapods-1.1.0.gem (100%)
Successfully installed cocoapods-1.1.0
Parsing documentation for cocoapods-1.1.0
Installing ri documentation for cocoapods-1.1.0
Done installing documentation for cocoapods after 2 seconds
1 gem installed
~/code/status-react (experiment) ☠  [2.4.1] [stable] % scripts/setup-osx
Installing requirements...

+ Updating to cocoapods 1.3.1
Fetching: cocoapods-1.3.1.gem (100%)
Successfully installed cocoapods-1.3.1
Parsing documentation for cocoapods-1.3.1
Installing ri documentation for cocoapods-1.3.1
Done installing documentation for cocoapods after 2 seconds
1 gem installed
```

andytudhope commented:
Perfect, works well, thank you. Would have got there in the end, but am busy fighting android studio because I couldn't find it in me to pay GenyMotion (or pay Apple for a Developer Certificate). Open source FTW! 🎉

dbalatero commented:
@andytudhope I am also fighting bootstrapping Android simulators. It's real crazy in there. Can you hit me up on Riot and maybe we can turn our findings into another script?

dbalatero commented:
@andytudhope @oskarth @yenda Is it ok to merge? If so I can move onto #2548. Thanks!

andytudhope commented:
I'm happy that this works and we can merge. @oskarth you around?

flexsurfer commented on scripts/setup-osx:
```diff
@@ -0,0 +1,56 @@
+dependency_setup lein deps
+dependency_setup npm install
+dependency_setup ./re-natal deps
+dependency_setup ./re-natal use-figwheel
```
these two
```
dependency_setup ./re-natal use-figwheel
dependency_setup lein re-frisk use-re-natal
``` 
are extra here
we need to run it each time after we run `./re-natal use-(ios/android)-device (simulator/genymotion)`

or we should add something by default, for example ios simulator

flexsurfer replied:
btw we could add leiningen aliases, something like 
```
:aliases {"status-genymotion" 
                     ["do"
                            ["shell" "./re-natal" "use-android-device" "genymotion"]
                            ["shell" "./re-natal" "use-figwheel"]
                            ["re-frisk" "use-re-natal"]]
```
and use it
`lein status-genymotion`

but implement in another PR

dbalatero commented:
@flexsurfer Removed those two lines. I agree about a separate PR. I've discussed getting some scripts together already to bootstrap simulators for iOS/Android, as well as re-opening your dev environment, which I think would be great to have!

pablodip commented on scripts/setup-osx:
```diff
@@ -0,0 +1,54 @@
+
+dependency_setup lein deps
+dependency_setup npm install
+dependency_setup ./re-natal deps
```
@dbalatero What is the difference between npm install and re-natal deps? I think they do the same (https://github.com/drapanjanas/re-natal/blob/master/re-natal.coffee#L144), and re-natal tries to use yarn if you have it installed, which I don't think it is ideal if the project doesn't use it yet. What do you think?

dbalatero replied:
Hey @pablodip! Totally agree that isn't ideal. re-natal _should_ check for `yarn.lock` before making those assumptions imho!

I'm ok with just doing `./re-natal deps`, so we should probably optimize for that.

I'm trying to get #2649 merged in, which will switch over to Yarn and make re-natal's weirdness a moot point.

Also I noticed that `@yarnpkg/lockfile` is marked as a dependency in the `develop` branch, which feels like a possible `git add . && git commit -m` goof? I wouldn't expect that package to be in the package-lock.json right now unless we were on yarn.

https://github.com/dbalatero/status-react/blob/c8168e9ff004e0bf523e27e039317b344030d9bf/package-lock.json#L7-L12

Anyways I think we should get #2649 in if it works out well, as it will speed up JS package installs significantly and has a more stable lock format in my experience. After that we can probably just switch to running `./re-natal deps` if it's already running `yarn install/npm install` like you said!

pablodip replied:
Thanks for replying @dbalatero! :)

Just doing `npm|yarn install` could be more explicit, fix the current bug, and depend less on external tooks. I am okay with waiting for that PR so long as it is ready soon. Otherwise it may be interesting to fix this issue even with npm since both the setup and run-osx scrips are broken for people who have installed yarn, which I believe is quite common.

As for `@yarnpkg/lockfile`, really no idea right now. :S

dbalatero replied:
@pablodip wait wait, is all that `./re-natal deps` does check if versions are ok (which the lock file should be used for in npm or yarn's case) and then run `npm/yarn install`? I'm currently observing this behavior.

If so we can just get rid of running that altogether, and just run the correct npm/yarn command, which I think is what you're suggesting :) **Edit:** and your suggestion took me this many comments to figure out, doh

dbalatero replied:
@pablodip Ok, I've removed that line on #2649's PR: https://github.com/status-im/status-react/pull/2649/files#diff-232884125f8738fe4f957f81e7915218L48

I'm just waiting to get that approved and merged. If you want to take any conversation over there to chat and/or encourage it to get merged that's great! Also if you have access to the private CI server, I need to know if my branch is passing or not heh.

pablodip replied:
@dbalatero 

What it does is just installing npm/yarn I believe: https://github.com/drapanjanas/re-natal/blob/master/re-natal.coffee#L952

Remember to remove that command on the `run-osx` script as well, please. :) (https://github.com/status-im/status-react/blob/develop/scripts/run-osx#L64)

I don't have access to the CI yet, and some people are probably on holidays. Will let them know though. :)
